Sources/LeafKit/LeafAST.swift

```diff
@@ -0,0 +1,128 @@
+    
+    let name: String
+    
+    private var rawAST: [Syntax]?
+    private var flatState: Bool?
+    
+    private(set) var ast: [Syntax]
+    private(set) var externalRefs = Set<String>()
+    private(set) var unresolvedRefs = Set<String>()
+    private(set) var flat: Bool
+    
+    init(name: String, ast: [Syntax]) {
+        self.name = name
+        self.ast = ast
+        self.rawAST = nil
+        self.flatState = nil
+        self.flat = false
+        
+        updateRefs()
+    }
+    
+     init(from: LeafAST, referencing externals: [String: LeafAST]) {
+        self.name = from.name
+        self.ast = from.ast
+        self.rawAST = from.rawAST
+        self.externalRefs = from.externalRefs
+        self.unresolvedRefs = from.unresolvedRefs
+        self.flatState = from.flatState
+        self.flat = from.flat
+        
+        guard self.flat == false else { return }
+        
+        inlineRefs(externals)
+    }
+    
+    mutating private func updateRefs() {
+        var firstRun = false
+        switch (flatState, rawAST) {
+            case (.some(true), _): return // known flat, no need to check for refs
+            case (.none, .some): fatalError("Invalid state: rawAST shouldn't hold a value if flatness isn't known")
+            case (.some(false), .none): fatalError("Invalid state: rawAST must hold a value if flatness is false")
+            case (.none, .none): firstRun = true
+            default: break
+        }
+        
+        unresolvedRefs.removeAll()
+        for syntax in ast {
+            switch syntax {
+                case .extend(let e): unresolvedRefs.insert(e.key)
+                default: break
+            }
+        }
+        flat = unresolvedRefs.count == 0
+        flatState = flat
+        
+        if firstRun == true {
+            if flat == false { rawAST = ast }
+            externalRefs = unresolvedRefs
+        }
+    }
+    
+    mutating private func inlineRefs(_ externals: [String: LeafAST]) {
+        guard externals.count > 0 else { return }
+        unresolvedRefs.removeAll()
+        var pos = ast.startIndex
+        
+        // inline provided externals
+        while pos < ast.endIndex {
+            if case .extend(let e) = ast[pos]  {
+                let key = e.key
+                if let insert = externals[key] {
+                    let inlined = e.extend(base: insert.ast)
+                    ast.remove(at: pos)
+                    ast.insert(contentsOf: inlined, at: pos)
+                    pos += inlined.count
+                } else {
+                    unresolvedRefs.insert(key)
+                }
+            }
+            pos += 1
+        }
+        
+        // compress raws
+        pos = ast.startIndex
+        while pos < ast.endIndex {
+            if case .raw(var syntax) = ast[pos], pos + 1 < ast.endIndex {
+                guard case .raw(var add) = ast[pos+1] else { pos += 1; break }
```

Comment:
Practically speaking, it's kind of a wash. Parsing a single template will always result in raws being efficient within the template, so there should never *BE* more than two RAWs in a row after inlining one external - at the boundaries of the original and inlined section.

So realistically I'd rather build the boundary compression of raws directly into inlining, but wanted it clear for the sake of review what was happening for now :)

Reply:
Makes sense - but I dare say (daring creature that I am 😆) that my version of the loop is still more efficient. Normally I'd call mine less readable, but I feel like in this case that much is _definitely_ a bit of a wash with all the math being done on the array index 😅

Reply:
Honestly, the main reason I'd suggest my version (or something like it) is it involves a lot less branching and maintenance of the index, meaning easier to read, easier to maintain, and easier to replace 🤷‍♀️

Reply:
From Discord: While I agree it can be more performant, should happen during inlining anyway, and for the moment I know this works and have aggressively profiled it so I don't want to screw with the monkey too much until the changes are easier to view on their own

Reply:
Inlining isn't _that_ good at what it does 😆 But yeah, that's fair.